pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -106,14 +106,24 @@ func NewGarbageCollector(
 		sharedInformers:  sharedInformers,
 		ignoredResources: ignoredResources,
 	}
-	if err := gb.monitorsForResources(deletableResources); err != nil {
-		return nil, err
+	if err := gb.syncMonitors(deletableResources); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to sync all monitors: %v", err))
 	}
 	gc.dependencyGraphBuilder = gb
 
 	return gc, nil
 }
 
+// SyncResourceMonitors starts or stops resource monitors as needed to ensure
+// that all (and only) those resources present in the map are monitored.
+func (gc *GarbageCollector) SyncResourceMonitors(deletableResources map[schema.GroupVersionResource]struct{}) error {
```

Comment:
Thanks for the review, I think we're now on the same page. That said, given this PR doesn't seem to be introducing the bug, should we address the worker refactor in a followup to keep the scope of this PR narrow? I'm happy to roll the changes into this PR if there's a need and if the increased scope is acceptable.